usr/sbin/p11sak/p11sak.c

```diff
@@ -938,50 +939,45 @@ static CK_RV tok_key_gen(CK_SESSION_HANDLE session, CK_ULONG keylength,
     /* Boolean attributes (cannot be specified by user) */
     CK_BBOOL a_token = ckb_true; // always true
     CK_BBOOL a_private = ckb_true; // always true
-
-    /* Boolean attributes from user input */
-    CK_BBOOL a_modifiable = ckb_false;
-    CK_BBOOL a_derive = ckb_false;
-    CK_BBOOL a_sensitive = ckb_false;
-    CK_BBOOL a_encrypt = ckb_false;
-    CK_BBOOL a_decrypt = ckb_false;
-    CK_BBOOL a_sign = ckb_false;
-    CK_BBOOL a_verify = ckb_false;
-    CK_BBOOL a_wrap = ckb_false;
-    CK_BBOOL a_unwrap = ckb_false;
-    CK_BBOOL a_extractable = ckb_false;
     CK_ULONG bs = sizeof(CK_BBOOL);
 
     /* Non-boolean attributes */
     CK_ULONG a_value_len = keylength / 8;
 
-    CK_ATTRIBUTE tmplt[] = {
-            // boolean attrs
-            { CKA_TOKEN, &a_token, bs }, 
-            { CKA_PRIVATE, &a_private, bs }, 
-            { CKA_MODIFIABLE, &a_modifiable, bs }, 
-            { CKA_DERIVE, &a_derive, bs },
-            { CKA_SENSITIVE, &a_sensitive, bs }, 
-            { CKA_ENCRYPT, &a_encrypt, bs }, 
-            { CKA_DECRYPT, &a_decrypt, bs }, 
-            { CKA_SIGN, &a_sign, bs }, 
-            { CKA_VERIFY, &a_verify, bs }, 
-            { CKA_WRAP, &a_wrap, bs }, 
-            { CKA_UNWRAP, &a_unwrap, bs },
-            { CKA_EXTRACTABLE, &a_extractable, bs },
-            // non-boolean attrs
-            { CKA_VALUE_LEN, &a_value_len, sizeof(CK_ULONG) }, 
-            { CKA_LABEL, label, strlen(label) } };
-    CK_ULONG num_attrs = sizeof(tmplt) / sizeof(CK_ATTRIBUTE);
-    CK_ULONG num_bools = num_attrs - 2;
-
-    /* set boolean attributes */
-    for (i = 0; i < (int) num_bools; i++) {
-        set_bool_attr_from_string(&tmplt[i], attr_string);
+    /* Standard template */
+    CK_ATTRIBUTE tmplt_std[] = {
+        { CKA_TOKEN, &a_token, bs }, 
+        { CKA_PRIVATE, &a_private, bs }, 
+        { CKA_VALUE_LEN, &a_value_len, sizeof(CK_ULONG) }, 
+        { CKA_LABEL, label, strlen(label) } 
+    };
+    CK_ULONG num_attrs = sizeof(tmplt_std) / sizeof(CK_ATTRIBUTE);
+
+    /* init template from String */
+    CK_ULONG num_attrs2;
+    if (attr_string) {
+        num_attrs2 = strlen(attr_string);
+    } else {
+        num_attrs2 = 0;
     }
+    CK_ATTRIBUTE tmplt[num_attrs2];
+
+    /* set boolean attributes, set template from string */
+    for (i = 0; i < (int) num_attrs2; i++) {
+        set_bool_attr_from_string(&tmplt[i], attr_string[i]);
+    }
+
+    /* build new template */
+    int mallocsize = (num_attrs + num_attrs2) * sizeof(CK_ATTRIBUTE);
+    CK_ATTRIBUTE* new_tmplt = malloc(mallocsize);
```

Comment:
Missing check for return value of `malloc`.  If `malloc` returns `NULL`, you could not allocate sufficient memory...

Reply:
If you anyway allocate a new template to hold both `tmplt_std` and `tmplt`, I do not see the need for `tmplt`.  Just allocate `new_tmplt`, run your loop for the boolean attributes, and copy the other template behind it (or the other way around if you want).

Reply:
Done